onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2832,210 @@ ONNX_OPERATOR_SET_SCHEMA(
+              // 2. Partial rotation: rotary embedding dimension is provided, rotary_embedding_dim = rotary_embedding_dim
+              builder.Add("HeadSize = Shape <start = 3, end = 4> (XReshaped)") // head_size
+                .Const1D("Two1D", (int64_t)2)
+                .Const1D("RotaryEmbedDimParam", rotary_embedding_dim)
+                .Const1D("Zero1D", (int64_t)0)
+                .Add("RotaryDimCond = Greater(RotaryEmbedDimParam, Zero1D)")
+                .Add("RotaryEmbedDim = Where(RotaryDimCond, RotaryEmbedDimParam, HeadSize)")
+                .Add("NoRotateLength = Sub(HeadSize, RotaryEmbedDim)")
+                .Add("RotateSplitLengths = Concat <axis = 0> (RotaryEmbedDim, NoRotateLength)");
+              // shape of input to rotate = input[:,:,:,:rotary_embedding_dim]
+              // shape of input not to rotate = input[:,:,:,rotary_embedding_dim:]
+              builder.Add("XToRotate, XNoRotate = Split <axis = -1, num_outputs = 2> (XReshaped, RotateSplitLengths)");
+
+              // Gather the cos and sine matrices from the respective caches using position ids.
+              // Unsqueeze applied to make cos and sin matrices have dimensions that are
+              // valid for multiplication with input when is split. For cases where rotary_embedding_dim is provided,
+              // slice the matrix values until that index only
+              builder.Add("CosCacheGather = Gather(cos_cache, position_ids)") // shape of cos matrix: [batch_size, sequence_length, head_size / 2]
+                .Add("SinCacheGather = Gather(sin_cache, position_ids)") // shape of cos matrix: [batch_size, sequence_length, head_size / 2]
+                .Add("RotaryEmbedDimHalf = Div(RotaryEmbedDim, Two1D)")
+                .Add("RotaryEmbedDimHalfInt = Cast (RotaryEmbedDimHalf)", "to", int_type)
+                .Add("CosCacheSliced = Slice(CosCacheGather, Zero1D, RotaryEmbedDimHalfInt, Two1D)") // shape of cos matrix: [batch_size, sequence_length, rotary_embedding_dim / 2]
+                .Add("SinCacheSliced = Slice(SinCacheGather, Zero1D, RotaryEmbedDimHalfInt, Two1D)") // shape of sin matrix: [batch_size, sequence_length, rotary_embedding_dim / 2]
+                .Add("CosCacheUnsqueezed = Unsqueeze <axes = [2]> (CosCacheSliced)") // shape of cos matrix: [batch_size, sequence_length, 1, rotary_embedding_dim / 2]
+                .Add("SinCacheUnsqueezed = Unsqueeze <axes = [2]> (SinCacheSliced)"); // shape of sin matrix: [batch_size, sequence_length, 1, rotary_embedding_dim / 2]
+
+              // Create slices of inputs to multiply with sin and cos matrices based on interleaved parameter
+              // For non-interleaved (basic) rotation, slices are created as follows,
+              builder.Add("X1Basic, X2Basic = Split <axis = -1, num_outputs = 2> (XToRotate)"); // shape of X1 = input[:,:,:,:rotary_embedding_dim/2], X2 = input[:,:,:,rotary_embedding_dim/2:rotary_embedding_dim]
```

Comment:
Can values be named with lower case, except for the inputs and outputs which should follow the convention https://onnx.ai/onnx/repo-docs/OpConventions.html

This way the values are not confused with the ops